2594 sensor back end link temperature logs to breaches

adrianwb opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2594

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->
The core of it is in two new functions - `get_sensor_logs_for_breach` (in `sensor\query.rs`) and `update_sensor_logs_for_breach` (in `sensor\update.rs`), both of which take a connection and a breach ID

- `get_sensor_logs_for_breach` returns all of the temperature logs which are considered to be part of the breach i.e. any which have temperatures outside the breach range, and which happen between the start & end of the breach. 
- `update_sensor_logs_for_breach` then uses `get_sensor_logs_for_breach` to return the matching temperature logs and assign them to the breach ID

Note that if it's a cumulative breach, that includes the whole day, as the start/end times for FridgeTag breaches (which are all cumulative) are not specified directly and have to be worked out, based only on the breach activation time (and the breach duration (cumulative time that day) outside the breach temperature range. That only works if the cumulative breach is actually consecutive i.e. if it doesn't have more than one continuous period of time during the day when the temperature was out of range. E.g. it could be out of range for an hour from 11am in the morning and then again for an hour from 4pm in the afternoon, in which case the duration of the breach is 2 hours, but the start time is 11am and the end time is 4pm => all temperature logs within that period are part of the breach if they're outside the temperature range. In that case, the start/end times of the breach are also extended to include the first & last temperature logs which fall outside the temperature range. 

# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->

I had a DB with existing sensor data, so to allow me to re-import the same Berlinger text file again, I used DB Browser to edit the SQL directly - I edited the `sensor` last connected date to before the first entry in the Berlinger text file, and set the end date of all existing breaches to null. The temperature_breach_id for temperature logs was already null, but if you're re-running a test, then you'll need to set those all to null as well.

(re) importing a Berlinger text file should then correctly assign breaches to temperature logs as described above.

TBC tomorrow...

## 💌 Any notes for the reviewer?
<!-- eg. Do you have any specific questions for the reviewer? Is there a high risk/complicated change they should focus on? If there are any general areas of the codebase your changes might have have touched or could cause side effects to, mention them here.-->

## 📃 Documentation
<!-- Note down any areas which require documentation updates -->
Import fridge tag docs _may_ need updated to include this, although the functionality already seems to work on the UI side once the breach IDs are assigned ;-) The only thing is that the displayed breach Duration is based on the difference between start/end timestamps, which isn't true for most cumulative breaches (see example above) - there is a `duration` field => better to use that instead => I'll submit a bug report tomorrow...
or

andreievg reviewed (commented):
Thanks Adrian. Overal logic seems sound, although I didn't actually look too deeply into re calculation of the breaches (or re-setting start and datetime for them), I think we may need to re-consider doing this. This is the main reason for not approving the PR yet. 

A few other things as per comments.

Also would be great to have unit tests, btw unit tests are also a great way to validate that the logic is correct (vs constatinly checking it manually) and to protect future dev work from regression and a great starting point for future dev work (i.e. if there is a bug, can easily add the use case in tests and just run that part of the code), here is an example (btw copied a lot from https://github.com/msupply-foundation/open-msupply/blob/cb8bb3121df2a062314715fb68d1de609325fe0b/server/repository/src/db_diesel/temperature_chart/temperature_chart.rs#L97):

https://github.com/msupply-foundation/open-msupply/compare/2594-Sensor-back-end---link-temperature-logs-to-breaches...2594-Sensor-back-end---link-temperature-logs-to-breaches(Unit-test-example)

Even though some time is needed to write the tests I've found that in the end it saves me time vs constant manual tests (and it become even fun to use the tests as a reference, to make them pass, sort of TDDish)

adrianwb commented on server/service/src/sensor/berlinger.rs:
```diff
@@ -61,13 +62,24 @@ pub fn get_matching_sensor_breach(
-        .start_datetime(DatetimeFilter::equal_to(start_datetime));
+        .r#type(EqualFilter::equal_to_breach_type(&breach_type));
+
+    match breach_type {
```
The start time on its own isn't guaranteed to be unique - you could have a consecutive and cumulative breach starting at the same time

andreievg replied:
But there is already a filter by breach_type, if it's filtered by both, wouldn't we find unique breach ?

adrianwb replied:
Yes,  start time + breach type should be unique (for a given sensor), but isn't that what this function is doing? The first bit is adding the sensor & breach type to the filter, and then the `match` is adding in the start time. The difference is that it's an exact match on the start time if it's a consecutive breach, but for cumulative breaches you can only have one per day so a start time any time during the same day is a match.

adrianwb replied:
As long as we never change the start time of a breach (which shouldn't happen if we're no longer trying to fix up slightly dodgy data), then the combination of start time and breach type (and sensor) should be unique. Thinking about time  zones makes my head hurt, but surely we're fine as long as all timestamps are adjusted so that any comparisons are all in the same time zone (or UTC or whatever)?

andreievg replied:
>  but surely we're fine as long as all timestamps are adjusted so that any comparisons are all in the same time zone (or UTC or whatever)?

Yep, likewise, makes my head hurt too. 

Agree if all timestamps are adjusted it should be easier to think about

adrianwb commented on server/service/src/sensor/query.rs:
```diff
@@ -37,3 +42,82 @@ pub fn get_sensor(ctx: &ServiceContext, id: String) -> Result<Sensor, SingleReco
+                desc: None,
+            };
+
+            match breach_record.r#type {
```
I'd noticed that as well, but I have a feeling that it may be more complicated - https://github.com/msupply-foundation/open-msupply/issues/2690#issuecomment-1885092383

andreievg replied:
Yeah there is definitely compilation with DLS, but I think that would happen regardless conversion to local time ?

adrianwb replied:
It's hard to see any way to nicely handle DST jumps - whether we stick with local time or change to UTC, the timestamps in the Berlinger files will be out by an hour for half the year (until such times as the Berlinger firmware is updated to take account of DST/TZ) => at least if we change to UTC, then it's independent of DST and it's in line with the rest of the DB. 

However... we do store the timestamp of when the sensor is initially activated, so that determines its (fixed) UTC offset => what if we always use that offset to convert to UTC, and then any UI display of the UTC timestamp will take account of DST and show the "real" local time?

andreievg replied:
> However... we do store the timestamp of when the sensor is initially activated

I wonder if we always know that ? Since we will be integrating berlinger data for existing sensors ? Hmm, for some reason i am think DST jump for berlinger data is not that important to capture, not sure why that is i think it's because:
* The frequency of berlinger sensor data integration is pretty low, and the precision of events is not as important as captured the existence of the events
* My previous leaning towards just getting the data into our system, as is, but matching our data modal, vs adjusting the data to be more 'accurate' then what's captured by fridgetag

adrianwb commented on server/service/src/sensor/query.rs:
```diff
@@ -37,3 +42,82 @@ pub fn get_sensor(ctx: &ServiceContext, id: String) -> Result<Sensor, SingleReco
+                match breach_record.r#type {
+                    TemperatureBreachRowType::ColdCumulative
+                    | TemperatureBreachRowType::ColdConsecutive => {
+                        if temperature_log.temperature_log_row.temperature
```
Hmm, that would be nice - I'll have a go and see...

adrianwb replied:
That looks good to me - I had gotten part way along the same path, but then got sucked into other things as often happens...

andreievg replied:
More then happy to try and help when that happens, i'll try to do examples like this, sorry that response is quite slow though

adrianwb commented on server/service/src/sensor/update.rs:
```diff
@@ -54,6 +60,98 @@ pub fn update_sensor(
+        }
+
+        for mut temperature_log in logs {
+            if let Some(_breach_id) = &temperature_log.temperature_log_row.temperature_breach_id {
```
The simple logic is that a re-import starts from the last saved timestamp of the previous import, so any temperature logs or breaches which have finished before that are ignored (and in any case, the import code won't update an existing temperature log or finished breach anyway) => no pre-existing records should re-sync.

Actually it's not quite that simple, as you could have had an ongoing breach at the time of the last saved timestamp - that's why the query for a unique breach doesn't include the end time, so that it will still get updated if the end time has changed.

FYI, the mSupply desktop code takes this a bit further to calculate consecutive breaches from the raw temperature log data - in that case, it rewinds to the end of the latest finished breach before the last saved timestamp (as you might not have had enough breached logs to trigger a breach at that point) => we'll have to do something similar if/when we do that in omSupply

andreievg replied:
> (and in any case, the import code won't update an existing temperature log or finished breach anyway

Do you mean currently or in a simpler logic ? To me, when i read this code, it looks like if there is a consecutive breach, we ignore if breach_id is set or not, and update the logs anyway ? (which would cause a re-sync of logs)

adrianwb replied:
It's been that way since the start - no existing records should get updated (or resynced) except for the case where there was an ongoing breach during the last import

adrianwb commented on server/service/src/sensor/update.rs:
```diff
@@ -54,6 +60,98 @@ pub fn update_sensor(
+
+                // Use the first log time if it's either earlier than the calculated breach end
+                // or if the calculated breach start is less than the sensor log interval before the first log time
+                if (breach_record.start_datetime > first_breach_datetime) | (log_interval > 0) & (breach_record.start_datetime < first_breach_datetime - Duration::seconds(log_interval.into())) {
```
I don't think it's an issue for Q-tags, but the raw data from Fridge tags doesn't actually have a breach end time (or a start time for that matter), just a duration and a trigger time, so the `temperature-sensor` crate tries to work it out. The problem is that this only works if it's a continuous cumulative breach (i.e. not split into more than one chunk during the day), so the calculated end time will _often_ be too early. As I discovered (see the testing section in #2594), the calculated end time can _sometimes_ also be too late. 

Hence it's more reliable to use the raw temperature log data, where it exists, to correct these miscalculations i.e. reset the breach start/end times when they aren't consistent with the first/last temperature logs 
I'm happy to create unit tests to validate that it all works.

That then leads to an additional adjustment to take account of the fact that temperature logs are only generated every X minutes, which means that you can easily end up with one cumulative breach having its last log at say 23:57 and the next cumulative breach having its first log at 00:02, when clearly it has all been continuous and the end/start time is actually midnight (there is an argument for combining these into a single cumulative breach which spans multiple days, but that's for the future).

I did consider doing these corrections inside the `temperature-sensor` crate, but it's much easier inside omSupply as all the data is in the DB.

andreievg replied:
> and a trigger time, so the temperature-sensor crate tries to work it out. The problem is that this only works if it's a continuous cumulative breach

Ahh true, I think we should remove the end date time for cumulative breaches returned by `temperature-sensor` crate (just to avoid confusion).

> Hence it's more reliable to use the raw temperature log data, where it exists, to correct these miscalculations

Oki, in that case I agree, but can't we just use first and last log for the day for this ? (as simple as we can make it ?). And maybe split this into a new method just for that rather then a side effect of `update_sensor_logs_for_breach` ?

> That then leads to an additional adjustment 
> (there is an argument for combining these into a single cumulative breach which spans multiple days, but that's for the future).

Cumulative breaches never made sense to me, you need a threshold duration **and** a boundary, but the boundary is never defined, for fridge tags it's 24 hours from midnight to midnight (which makes little sense, since you can have a cumulative breach at night, but it won't be identified, and actually this is a likely scenario for power outage or for fridge unplugged, since no one is at the facility to fix it), in cold chain app the boundary is whatever 'dataset' you are looking at (the chart to and from date). What I am trying to get at here, is that it's a guessing game of 'what is the right' way to do cumulative breach, and we should just take fridge tag definition of it. (and not worry about special cases like continuous cumulative past midnight).

Saying all that, how does this sound:

For cumulative, use last log of the day that matches breach threshold, and set start and end date for this ? (start date should be know right ? even if it's not accurate, I think we should use it, to keep consistent with fridge tag data).

And of course associate every temperature log on the day that matches breach threshold with that cumulative breach (but of course do the consecutive first, and filter out not null breach ids for cumulative)

adrianwb replied:
> > and a trigger time, so the temperature-sensor crate tries to work it out. The problem is that this only works if it's a continuous cumulative breach
> 
> Ahh true, I think we should remove the end date time for cumulative breaches returned by `temperature-sensor` crate (just to avoid confusion).

I'm not quite sure that I follow?

> > Hence it's more reliable to use the raw temperature log data, where it exists, to correct these miscalculations
> 
> Oki, in that case I agree, but can't we just use first and last log for the day for this ? (as simple as we can make it ?). 

Even if we just use the temperature logs, then I think we should still do the midnight rounding (first two steps in https://github.com/msupply-foundation/open-msupply/pull/2629#discussion_r1488237477). 

Either way, the downside of using just the temperature logs is that you end up editing the start & end of _all_ the breaches, not just the few exceptions which have dodgy data, and you potentially discard some relevant data e.g. if you have an actual breach from 14:30-15:01 (say you have a threshold duration of 30 minutes), with breached logs at 14:32, 14:37 .. 14:57, then it appears that the breach was only 25 minutes (i.e. less than the threshold duration). Having said that, the stored breach duration will still be 31 minutes - that is independent of the start/end timestamps for cumulative breaches.

> And maybe split this into a new method just for that rather then a side effect of `update_sensor_logs_for_breach` ?

Fair enough - I can rename this method to something like `assign_sensor_logs_to_breach` and split the extra logic out into something like `update_breach_from_sensor_logs`

> > That then leads to an additional adjustment
> > (there is an argument for combining these into a single cumulative breach which spans multiple days, but that's for the future).
> 
> Cumulative breaches never made sense to me, you need a threshold duration **and** a boundary, but the boundary is never defined, for fridge tags it's 24 hours from midnight to midnight (which makes little sense, since you can have a cumulative breach at night, but it won't be identified, and actually this is a likely scenario for power outage or for fridge unplugged, since no one is at the facility to fix it), in cold chain app the boundary is whatever 'dataset' you are looking at (the chart to and from date). What I am trying to get at here, is that it's a guessing game of 'what is the right' way to do cumulative breach, and we should just take fridge tag definition of it. (and not worry about special cases like continuous cumulative past midnight).

I absolutely agree on the whole "how do you define a cumulative breach?" thing - for Berlinger tags, the inference from the data is that it's based on a 24-hour midnight to midnight period, but that's only for them and I can see that there's a _very_ good argument for working it out on demand for any given interval. However, I don't think the fridge breakdown overnight example works - unless the fridge fixes itself very soon after midnight, you're going to end up with two cumulative breaches, one for each day, and that's going to be noticed.

> Saying all that, how does this sound:
> 
> For cumulative, use last log of the day that matches breach threshold, and set start and end date for this ? (start date should be know right ? even if it's not accurate, I think we should use it, to keep consistent with fridge tag data).

Of course, the main point of fridge monitors is that they detect and flag breaches, and it really doesn't make much practical difference if we're a few minutes out either way. It just pains me to leave something "nearly right most of the time" when I can make it "right all of the time" with just a little bit of extra work!

> And of course associate every temperature log on the day that matches breach threshold with that cumulative breach (but of course do the consecutive first, and filter out not null breach ids for cumulative)

Absolutely, that's the core of this PR

adrianwb replied:
Fair enough - lets just leave it as the bare minimum "assume that the sensor data can be trusted" for now anyway. I'd still be keen to add in some (probably fairly extensive) comments in the code to explain the potential failings in the raw data, while I still have the context (sort-of) loaded - although maybe the best place for that is in the docs for the `temperature-sensor` crate?

andreievg replied:
>  although maybe the best place for that is in the docs for the temperature-sensor crate?

Yeah I think that would be great, thank you !

adrianwb commented:
> Also would be great to have unit tests, btw unit tests are also a great way to validate that the logic is correct (vs constatinly checking it manually) and to protect future dev work from regression and a great starting point for future dev work (i.e. if there is a bug, can easily add the use case in tests and just run that part of the code), here is an example (btw copied a lot from
> 
> https://github.com/msupply-foundation/open-msupply/blob/cb8bb3121df2a062314715fb68d1de609325fe0b/server/repository/src/db_diesel/temperature_chart/temperature_chart.rs#L97
> 
> ):
> [2594-Sensor-back-end---link-temperature-logs-to-breaches...2594-Sensor-back-end---link-temperature-logs-to-breaches(Unit-test-example)](https://github.com/msupply-foundation/open-msupply/compare/2594-Sensor-back-end---link-temperature-logs-to-breaches...2594-Sensor-back-end---link-temperature-logs-to-breaches(Unit-test-example))
> 
> Even though some time is needed to write the tests I've found that in the end it saves me time vs constant manual tests (and it become even fun to use the tests as a reference, to make them pass, sort of TDDish)

Yes, I'll do that - just need to generate some test sensor files...

andreievg commented on server/service/src/sensor/update.rs:
```diff
@@ -54,6 +60,98 @@ pub fn update_sensor(
+                    TemperatureBreachRowRepository::new(connection).upsert_one(&breach_record)?;
+                }
+            }
+            if let Some(last_log) = logs.last() {
```
> and start

Sorry I missed that, isn't start date always provided under `A:` ?

adrianwb replied:
The breach start and end timestamps _are_ provided for Q-tags, so those are fine (and none of these adjustments are needed as we can just use the provided breach data as is), but not for Fridge-tags - all we get is the breach activation timestamp (which is when the breach has been going long enough to trigger when it gets to the breach threshold duration) and the breach duration, so the best we can do is to try to work out the start & end from the data that we do have:

(1) for a fridge tag without logging, we have the activation timestamp and breach duration, and a few implied temperature logs for the min/max temperatures for each day:
- the crate calculates the cumulative breach start timestamp by subtracting the breach threshold duration from the activation timestamp (or midnight if the activation timestamp is within the breach threshold duration from midnight, which indicates an ongoing breach from the previous day) [as mentioned above, this only works if there are no gaps in the cumulative breach up to the activation timestamp]
- the crate calculates the cumulative breach end timestamp by adding on the breach duration to the calculated start timestamp (or midnight if that is sooner) [and this only works if there are no gaps in the cumulative breach at all, although you could also work it out by adding the (breach duration - breach threshold duration) to the activation timestamp, and that would maybe allow for a gap before the activation timestamp]
-> there's nothing more that omSupply can do, other than assign any of the min/max temperature logs to a cumulative breach if they happen to fall within its duration
=> there are minimal temperature logs available (just daily min/max); cumulative breach start/end timestamps are only valid if it's a continuous breach; no consecutive breach data unless you say that all cumulative breaches are also consecutive ones (which may be a valid approach to take??). All of this may be moot as I understand that we're not committing to support this type of fridge tag anyway @adamdewey ?

(2) for fridge tags with logging data, we have additional temperature logs, usually at 5-minute intervals. That lets us cross-check the calculated breach data for consistency - you'd expect that any hot/cold breach temperature logs should be within the corresponding breach start/end timestamp range, and that's when you discover the discrepancies, especially if there's a gap (bearing in mind that we only get _cumulative_ breaches from the Fridge-tag data)...
-> we can use the temperature logs to correct the limitations in the breach start/end calculations:
- if the first breached temperature log is within the logging interval to midnight, then use midnight as the effective log timestamp
- if the last breached temperature log is within the logging interval from midnight, then use midnight as the effective log timestamp in the adjustments below
- if the first breached temperature log is before the breach start timestamp, then extend the breach start range to include the first log timestamp in the adjustments below
- if the last breached temperature log is after the breach end timestamp, then extend the breach range to include the last log timestamp
- if the first breached temperature log is more than the logging interval after the breach start timestamp, then reduce the breach range to start with the first log timestamp
- if the last breached temperature log is more than the logging interval before the breach end timestamp, then reduce the breach range to end with the last log timestamp
=> the first two of these are to avoid the (completely naff IMHO) situation where, for example, an all-day cumulative breach is shown as having started at 00:02 and ended at 23:57, due to the 5-minute logging interval; the others correct discrepancies in the calculated breach start/end timestamps; if all temperature logs are within the given breach range, then nothing gets changed

adrianwb replied:
To put your mind at rest, the crate itself doesn't attempt to do any data patching up - the only thing it does is to try to work out the start times of cumulative breaches for fridgetags from the breach activation time minus the breach config duration, and then adds on the breach duration to get the breach end time => those won't be accurate in some cases, but the most important thing is the breach duration itself, not the exact times when it started & finished.

I don't remember off the top of my head if that info is already in the crate docs, but I'll add it if it isn't.

andreievg commented:
> Yes, I'll do that - just need to generate some test sensor files...

As per example, we don't actually need a test sensor file, we can mock that data, I think that would be easier to read, test and modify in the future

adrianwb commented on server/service/src/sensor/query.rs:
```diff
@@ -37,3 +42,82 @@ pub fn get_sensor(ctx: &ServiceContext, id: String) -> Result<Sensor, SingleReco
+                | TemperatureBreachRowType::HotCumulative => {
+                    // Cumulative breach can include any time on the same day (can only be at most one of hot/cold starting per day)
+                    let zero_time = NaiveTime::parse_from_str("00:00", "%H:%M").unwrap(); // hard-coded -> should always work!
+                    let start_breach = NaiveDateTime::new(breach_record.start_datetime.date(), zero_time); // set to start of day
```
`and_hms` is deprecated, but `and_hms_opt(0,0,0).unwrap()` works - https://github.com/msupply-foundation/open-msupply/pull/2629/commits/91ce1882eba15fe7c6e2bf5595c072a99d418ac4

andreievg reviewed (commented):
@adrianwb, sorry it took so long to re-review this PR. I've made quite a few comments. I wanted to see how the code looks and behaves if we just assume that the data is correct for breaches and simply associate temperature logs with breaches based on start and end datetime of the breach (with consequitive breach taking the priority), and also introduce basic tests to validate that the logic is correct. It ended up being quite a few changes so I've made a PR, can you please check it out: https://github.com/msupply-foundation/open-msupply/pull/3443

To be honest, I think that the extra code for correcting start and end time of fridge tag breaches would be hard to maintain in the future. And that's the main reason why I strugle to give this PR the approval (and made another PR with simplified logic), even though I fully trust your implementation and the analysis of fridge tag data to deduce the business logic for those adjustments

adrianwb commented:
> @adrianwb, sorry it took so long to re-review this PR. I've made quite a few comments. I wanted to see how the code looks and behaves if we just assume that the data is correct for breaches and simply associate temperature logs with breaches based on start and end datetime of the breach (with consequitive breach taking the priority), and also introduce basic tests to validate that the logic is correct. It ended up being quite a few changes so I've made a PR, can you please check it out: #3443
> 
> To be honest, I think that the extra code for correcting start and end time of fridge tag breaches would be hard to maintain in the future. And that's the main reason why I strugle to give this PR the approval (and made another PR with simplified logic), even though I fully trust your implementation and the analysis of fridge tag data to deduce the business logic for those adjustments

I'll have a look at #3443 tomorrow and do some tests with my set of import files - all being well, I'll get it merged into the parent branch before too long

andreievg commented:
> I'll have a look at https://github.com/msupply-foundation/open-msupply/pull/3443 tomorrow and do some tests with my set of import files - all being well, I'll get it merged into the parent branch before too long

Thanks @adrianwb, would be great to get your take on it. I'll try to get back to you a bit quicker, to get this PR over the line

andreievg commented:
@adrianwb, it would be great to get this one in the next release, I think the PR merge deadline is next Wednesday (24th of April). Please let us know if you don't quite have the time, will try to arrange review from someone in NZ office. Thanks !

adrianwb commented:
> @adrianwb, it would be great to get this one in the next release, I think the PR merge deadline is next Wednesday (24th of April). Please let us know if you don't quite have the time, will try to arrange review from someone in NZ office. Thanks !

I'm planning to get this done and dusted tomorrow or possibly Friday, before I'm off to sunny Portugal for a week :-)

roxy-dao commented:
> > @adrianwb, it would be great to get this one in the next release, I think the PR merge deadline is next Wednesday (24th of April). Please let us know if you don't quite have the time, will try to arrange review from someone in NZ office. Thanks !
> 
> I'm planning to get this done and dusted tomorrow or possibly Friday, before I'm off to sunny Portugal for a week :-)

Awesome~ Can you also update with v2.0.0-RC branch and change base branch (: Thanks!

adrianwb commented:
> > > @adrianwb, it would be great to get this one in the next release, I think the PR merge deadline is next Wednesday (24th of April). Please let us know if you don't quite have the time, will try to arrange review from someone in NZ office. Thanks !
> > 
> > 
> > I'm planning to get this done and dusted tomorrow or possibly Friday, before I'm off to sunny Portugal for a week :-)
> 
> Awesome~ Can you also update with v2.0.0-RC branch and change base branch (: Thanks!

Done, and I've fixed a couple of compilation issues so that it all compiles now and all cargo tests are passing. I'm just getting several macro warnings for some files:

![image](https://github.com/msupply-foundation/open-msupply/assets/6196269/60458520-ea97-4109-9f62-e04985c84beb)

I'd thought it might have been my older Rust version, but it's still happening after updating to the latest 1.77.2 and a cargo clean - hopefully it's something that someone else has already found and fixed elsewhere...

I ran out of time to do some real-life tests with my Berlinger data, but will try to find a bit of time to do that early next week, but feel free to make any other changes as needed so that you can get it over the line (finally!).

I said that I'd add some explanatory docs to the `temperature-sensor` crate, so I'll do that when I get back, but that can be a separate issue - no need to hold this back on that account.

andreievg commented:
Thanks @adrianwb for all the work on this, i just did a compile and test, seems to be working
